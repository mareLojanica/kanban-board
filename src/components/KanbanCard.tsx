import { CSS, Transform } from "@dnd-kit/utilities"
import { useDraggable } from "@dnd-kit/core"
import { useDispatch } from "react-redux"
import { useCallback, useState } from "react"
import {
	Box,
	IconButton,
	TextField,
	Typography,
	useMediaQuery,
	useTheme,
} from "@mui/material"
import CloseIcon from "@mui/icons-material/Close"

import {
	deleteSelectedTicket,
	updateTicketById,
} from "../features/kanban/ticketSlice"
import { KanbanCardProps } from "../types/VerticalKanbanBoard.types"
import {
	useDeleteTicketMutation,
	useUpdateTicketMutation,
} from "../api/tickets.grapql"
import { styled } from "@mui/system"

const StyledCard = styled(Box, {
	shouldForwardProp: (prop) =>
		prop !== "transform" &&
		prop !== "isDragging" &&
		prop !== "isMobileOrTablet" &&
		prop !== "cardColor",
})<{
	transform: Transform
	isMobileOrTablet: boolean
	isDragging: boolean
	cardColor: string
}>(({ theme, transform, isMobileOrTablet, isDragging, cardColor }) => ({
	display: "flex",
	alignItems: "center",
	justifyContent: "space-between",
	padding: theme.spacing(2),
	backgroundColor: cardColor,
	marginBottom: theme.spacing(2),
	transform: CSS.Translate.toString(transform),
	opacity: isDragging ? 0.5 : 1,
	cursor: "grab",
	zIndex: 100,
	height: "80px",
	boxShadow: isMobileOrTablet ? "0px 0px 8px 4px rgba(0, 0, 0, 0.2)" : "none",
	"&:hover .iconButton": {
		display: "block",
		transform: "scale(1.1)",
	},
}))

const KanbanCard = ({
	id,
	title = "New Ticket",
	index,
	parent,
	isNew = false,
	cardColor,
}: KanbanCardProps) => {
	const dispatch = useDispatch()
	const [deleteTicket] = useDeleteTicketMutation()
	const [updateTicket] = useUpdateTicketMutation()
	const [isEditing, setIsEditing] = useState(isNew)
	const [editedTitle, setEditedTitle] = useState(isNew ? "" : title)

	const theme = useTheme()
	const isMobileOrTablet = useMediaQuery(theme.breakpoints.down("md"))

	const { attributes, listeners, setNodeRef, transform, isDragging } =
		useDraggable({
			id,
			data: { id, title, index, parent },
		})

	const handleDelete = useCallback(async () => {
		try {
			await deleteTicket({ id }).unwrap()
			dispatch(deleteSelectedTicket(id))
		} catch (error) {
			console.error("Error deleting ticket:", error)
		}
	}, [id, deleteTicket, dispatch])

	const handleTitleEdit = useCallback(async () => {
		if (editedTitle.trim() === title) return
		setIsEditing(false)
		const updatedTicket = { id, title: editedTitle }

		try {
			await updateTicket(updatedTicket).unwrap()
			dispatch(updateTicketById(updatedTicket))
		} catch (error) {
			console.error("Error updating ticket:", error)
		}
	}, [id, editedTitle, title, updateTicket, dispatch])

	const handleInputKeyDown = useCallback(
		(e: React.KeyboardEvent) => {
			if (e.key === "Enter") handleTitleEdit()
			if (e.key === "Escape") setIsEditing(false)
		},
		[handleTitleEdit]
	)

	return (
		<StyledCard
			ref={setNodeRef}
			{...listeners}
			{...attributes}
			cardColor={cardColor}
			transform={transform}
			isDragging={!!isDragging}
			isMobileOrTablet={isMobileOrTablet}
			onDoubleClick={() => setIsEditing(true)}
			data-testid={`kanban-card-${id}`}
		>
			{isEditing ? (
				<TextField
					value={editedTitle}
					onChange={(e) => setEditedTitle(e.target.value)}
					onBlur={handleTitleEdit}
					onKeyDown={handleInputKeyDown}
					id="edit-input-id"
					autoFocus
					variant="standard"
					fullWidth
					sx={{
						flexGrow: 1,
						marginRight: 2,
						input: { color: "#fff" },
					}}
				/>
			) : (
				<Typography
					variant="body1"
					sx={{ cursor: "text", color: "#fff", flexGrow: 1 }}
				>
					{editedTitle}
				</Typography>
			)}

			<IconButton
				onClick={(e) => {
					e.stopPropagation()
					handleDelete()
				}}
				data-testid="delete-ticket"
				className="iconButton"
				sx={{
					p: 2,
					display: isMobileOrTablet ? "block" : "none",
					color: "#fff",
				}}
			>
				<CloseIcon fontSize="small" />
			</IconButton>
		</StyledCard>
	)
}

export default KanbanCard
