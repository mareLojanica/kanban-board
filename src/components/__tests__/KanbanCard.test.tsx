import { render, fireEvent, screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import { Provider } from "react-redux"
import configureStore from "redux-mock-store"
import { act } from "react"

import {
	useDeleteTicketMutation,
	useUpdateTicketMutation,
} from "../../api/tickets.grapql"
import KanbanCard from "../KanbanCard"

jest.mock("../../api/tickets.grapql", () => ({
	useDeleteTicketMutation: jest.fn(),
	useUpdateTicketMutation: jest.fn(),
}))

const mockStore = configureStore([])

describe("KanbanCard Component", () => {
	let store
	const mockDeleteTicket = jest.fn(() => ({
		unwrap: jest.fn().mockResolvedValue({}),
	}))
	const mockUpdateTicket = jest.fn(() => ({
		unwrap: jest.fn().mockResolvedValue({}),
	}))

	;(useDeleteTicketMutation as jest.Mock).mockReturnValue([mockDeleteTicket])
	;(useUpdateTicketMutation as jest.Mock).mockReturnValue([mockUpdateTicket])

	beforeEach(() => {
		jest.clearAllMocks()
		store = mockStore({ tickets: {} })
	})

	const renderComponent = (props = {}) =>
		render(
			<Provider store={store}>
				<KanbanCard
					id="1"
					title="Test Ticket"
					index={0}
					parent="todo"
					cardColor="#fff"
					{...props}
				/>
			</Provider>
		)

	it("renders the KanbanCard correctly", () => {
		const { container } = renderComponent()
		expect(container).toMatchSnapshot()
	})

	it("allows editing the title", async () => {
		const { container } = renderComponent()
		const user = userEvent.setup()

		const titleElement = screen.getByText("Test Ticket")
		fireEvent.doubleClick(titleElement)

		const input = container.querySelector("#edit-input-id")
		expect(input).toBeInTheDocument()

		await user.clear(input)
		await user.type(input, "Updated")

		fireEvent.blur(input)

		expect(mockUpdateTicket).toHaveBeenCalledWith({
			id: "1",
			title: "Updated",
		})
	})

	it("triggers delete action when delete button is clicked", async () => {
		renderComponent()

		const user = userEvent.setup()
		const deleteButton = screen.getByTestId("delete-ticket")

		await user.click(deleteButton)

		expect(mockDeleteTicket).toHaveBeenCalledWith({ id: "1" })
	})

	it("applies drag styles correctly when dragging", () => {
		renderComponent()
		const cardElement = screen.getByText("Test Ticket").closest("div")
		expect(cardElement).toHaveStyle("cursor: grab")
	})
})

describe("KanbanCard Component - Redux Integration", () => {
	let store
	const mockDeleteTicket = jest.fn(() => ({
		unwrap: jest.fn().mockResolvedValue({}),
	}))
	const mockUpdateTicket = jest.fn(() => ({
		unwrap: jest.fn().mockResolvedValue({}),
	}))

	beforeEach(() => {
		jest.clearAllMocks()
		store = mockStore({ tickets: {} })
		;(useDeleteTicketMutation as jest.Mock).mockReturnValue([
			mockDeleteTicket,
		])
		;(useUpdateTicketMutation as jest.Mock).mockReturnValue([
			mockUpdateTicket,
		])
	})

	const renderComponent = (props = {}) =>
		render(
			<Provider store={store}>
				<KanbanCard
					id="1"
					title="Test Ticket"
					index={0}
					parent="todo"
					cardColor="#fff"
					{...props}
				/>
			</Provider>
		)

	it("dispatches deleteSelectedTicket when delete button is clicked", async () => {
		renderComponent()

		const user = userEvent.setup()
		const deleteButton = screen.getByTestId("delete-ticket")

		await user.click(deleteButton)

		expect(mockDeleteTicket).toHaveBeenCalledWith({ id: "1" })

		const actions = store.getActions()
		expect(actions).toEqual(
			expect.arrayContaining([
				expect.objectContaining({
					type: "tickets/deleteSelectedTicket",
					payload: "1",
				}),
			])
		)
	})

	it("dispatches updateTicketById when title is edited", async () => {
		const { container } = renderComponent()
		const user = userEvent.setup()

		const titleElement = screen.getByText("Test Ticket")
		await act(async () => {
			fireEvent.doubleClick(titleElement)
		})

		const input = container.querySelector(
			"#edit-input-id"
		) as HTMLInputElement
		expect(input).toBeInTheDocument()

		await user.clear(input)
		await user.type(input, "Updated")

		await act(async () => {
			input.blur()
		})

		const actions = store.getActions()

		expect(mockUpdateTicket).toHaveBeenCalledWith({
			id: "1",
			title: "Updated",
		})

		expect(actions).toEqual(
			expect.arrayContaining([
				expect.objectContaining({
					type: "tickets/updateTicketById",
					payload: { id: "1", title: "Updated" },
				}),
			])
		)
	})
})
