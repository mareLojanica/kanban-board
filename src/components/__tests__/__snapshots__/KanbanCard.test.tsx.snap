// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KanbanCard Component renders the KanbanCard correctly 1`] = `
<div>
  <div
    aria-describedby=""
    aria-disabled="false"
    aria-roledescription="draggable"
    class="MuiBox-root css-1ax3bw2"
    data-testid="kanban-card-1"
    role="button"
    tabindex="0"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 css-1o5jgl7-MuiTypography-root"
    >
      Test Ticket
    </p>
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium iconButton css-15lzfo0-MuiButtonBase-root-MuiIconButton-root"
      data-testid="delete-ticket"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-120dh41-MuiSvgIcon-root"
        data-testid="CloseIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </button>
  </div>
</div>
`;
